From 15f0ee0cdf2e74c23f6756599a3983557763082e Mon Sep 17 00:00:00 2001
Message-Id: <15f0ee0cdf2e74c23f6756599a3983557763082e.1413837857.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: Durgadoss R <durgadoss.r@intel.com>
Date: Wed, 3 Sep 2014 11:54:10 +0530
Subject: [PATCH 308/312] FOR_UPSTREAM [VPG]: drm/i915: Program PFI credits

PFI stands for Pondicherry Fabric Interface.
PFI credits can be assigned to Pondicherry memory arbiter
to get better performance. This patch adds support to
program PFI credits based on cdclk and czclk ratio. Highest
credits are programmed to get the most performance.

Issue: GMINL-1042
Change-Id: Ic4218da44f80b5df41b4f5b94318835abfd2aa05
Signed-off-by: Gajanan Bhat <gajanan.bhat@intel.com>
Signed-off-by: Durgadoss R <durgadoss.r@intel.com>
---
 drivers/gpu/drm/i915/i915_reg.h |    6 ++++++
 drivers/gpu/drm/i915/intel_pm.c |   11 ++++++++---
 2 files changed, 14 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_reg.h b/drivers/gpu/drm/i915/i915_reg.h
index 733da97..f1312c3 100644
--- a/drivers/gpu/drm/i915/i915_reg.h
+++ b/drivers/gpu/drm/i915/i915_reg.h
@@ -432,6 +432,12 @@
 #define SRC_COPY_BLT  ((0x2<<29)|(0x43<<22))
 
 #define GCI_CONTROL	(dev_priv->info.display_mmio_offset + 0x650C)
+#define VGA_FAST_MODE_DISABLE	(1<<14)
+#define PFI_CREDITS_15		(7<<28)
+#define PFI_CREDITS_31		(8<<28)
+#define PFI_CREDITS_63		(9<<28)
+#define FORCE_CREDIT_RESEND	(1<<27)
+
 /*
  * Registers used only by the command parser
  */
diff --git a/drivers/gpu/drm/i915/intel_pm.c b/drivers/gpu/drm/i915/intel_pm.c
index 047ac36..e699dfe 100644
--- a/drivers/gpu/drm/i915/intel_pm.c
+++ b/drivers/gpu/drm/i915/intel_pm.c
@@ -7977,18 +7977,23 @@ int vlv_freq_opcode(struct drm_i915_private *dev_priv, int val)
 void program_pfi_credits(struct drm_i915_private *dev_priv, bool flag)
 {
 	int cd_clk, cz_clk;
+	int val = VGA_FAST_MODE_DISABLE | FORCE_CREDIT_RESEND;
 
 	if (!flag) {
-		I915_WRITE(GCI_CONTROL, 0x00004000);
+		I915_WRITE(GCI_CONTROL, VGA_FAST_MODE_DISABLE);
 		return;
 	}
 
 	intel_get_cd_cz_clk(dev_priv, &cd_clk, &cz_clk);
 
 	if (cd_clk >= cz_clk) {
+		if (IS_CHERRYVIEW(dev_priv->dev))
+			val |= PFI_CREDITS_63;
+		else if (IS_VALLEYVIEW(dev_priv->dev))
+			val |= PFI_CREDITS_15;
 		/* Disable before enabling */
-		I915_WRITE(GCI_CONTROL, 0x00004000);
-		I915_WRITE(GCI_CONTROL, 0x78004000);
+		I915_WRITE(GCI_CONTROL, VGA_FAST_MODE_DISABLE);
+		I915_WRITE(GCI_CONTROL, val);
 	} else
 		DRM_ERROR("cd clk < cz clk");
 }
-- 
1.7.9.5

